### PR TITLE
feat: Kafka external streams read from specific partitions

### LINKED ISSUE #201: Partial partitions consumption for external stream
**Use case**

As a user, i would like to consume / live query specific topic partitions for external Kafka stream.

**Describe the solution you'd like**

```
SELECT * FROM my_kstream SETTINGS shards=1,3;
```

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by zliang-min**
Closes #201 

PR checklist:
- Did you run ClangFormat ?
- Did you separate headers to a different section in existing community code base ?
- Did you surround `proton: starts/ends` for new code in existing community code base ?

Please write user-readable short description of the changes:

### CODE REVIEW

File: src/KafkaLog/KafkaWAL.cpp [hunk 1]
```diff
@@ -800,8 +800,17 @@ KafkaWALClusterPtr KafkaWAL::cluster(const KafkaWALContext & ctx) const
...
+std::vector<int64_t> KafkaWAL::offsetsForTimestamps(const std::string & topic, const std::vector<PartitionTimestampPair> & partitionTimestamps, int32_t timeout_ms) const
+{
+    return getOffsetsForTimestamps(producer_handle.get(), topic, partitionTimestamps, timeout_ms);
+}
+
 std::vector<int64_t> KafkaWAL::offsetsForTimestamps(const std::string & topic, const std::vector<int64_t> & timestamps, int32_t timeout_ms) const
 {
-    return getOffsetsForTimestamps(producer_handle.get(), topic, timestamps, timeout_ms);
+    std::vector<PartitionTimestampPair> partitionTimestamps{timestamps.size()};
+    for (size_t i{0}; i < timestamps.size(); ++i)
```
Comment (chenziliang) [inline src/KafkaLog/KafkaWAL.cpp]: for(size_t i = 0; auto t : timestamps)
  partition_timestamps.emplace_back(i++, t);

File: src/KafkaLog/KafkaWAL.cpp [hunk 2]
```diff
@@ -800,8 +800,17 @@ KafkaWALClusterPtr KafkaWAL::cluster(const KafkaWALContext & ctx) const
...
 /// https://cwiki.apache.org/confluence/display/KAFKA/KIP-33+-+Add+a+time+based+log+index
+std::vector<int64_t> KafkaWAL::offsetsForTimestamps(const std::string & topic, const std::vector<PartitionTimestampPair> & partitionTimestamps, int32_t timeout_ms) const
+{
+    return getOffsetsForTimestamps(producer_handle.get(), topic, partitionTimestamps, timeout_ms);
+}
+
 std::vector<int64_t> KafkaWAL::offsetsForTimestamps(const std::string & topic, const std::vector<int64_t> & timestamps, int32_t timeout_ms) const
 {
-    return getOffsetsForTimestamps(producer_handle.get(), topic, timestamps, timeout_ms);
+    std::vector<PartitionTimestampPair> partitionTimestamps{timestamps.size()};
```
Comment (chenziliang) [inline src/KafkaLog/KafkaWAL.cpp]: "partition_timestamps;"
partition_timestamps.reserve(...)

File: src/KafkaLog/KafkaWALCommon.cpp [hunk 3]
```diff
@@ -232,40 +232,42 @@ DescribeResult describeTopic(const String & name, struct rd_kafka_s * rk, Poco::
     return {.err = DB::ErrorCodes::RESOURCE_NOT_FOUND};
 }
 
-std::vector<int64_t>
-getOffsetsForTimestamps(struct rd_kafka_s * rd_handle, const std::string & topic, const std::vector<int64_t> & timestamps, int32_t timeout_ms)
+std::vector<int64_t> getOffsetsForTimestamps(
+    struct rd_kafka_s * rd_handle,
+    const std::string & topic,
+    const std::vector<PartitionTimestampPair> & partitionTimestamps,
```
Comment (chenziliang) [inline src/KafkaLog/KafkaWALCommon.cpp]: Let's pay attention to the name conventions of a variable.

File: src/KafkaLog/KafkaWALCommon.cpp [hunk 4]
```diff
@@ -232,40 +232,42 @@ DescribeResult describeTopic(const String & name, struct rd_kafka_s * rk, Poco::
...
     assert(rd_handle);
 
     using RdKafkaTopicPartitionListPtr
         = std::unique_ptr<rd_kafka_topic_partition_list_t, decltype(rd_kafka_topic_partition_list_destroy) *>;
-    RdKafkaTopicPartitionListPtr offsets{rd_kafka_topic_partition_list_new(static_cast<int>(timestamps.size())), rd_kafka_topic_partition_list_destroy};
+    RdKafkaTopicPartitionListPtr offsets{
+        rd_kafka_topic_partition_list_new(static_cast<int>(partitionTimestamps.size())), rd_kafka_topic_partition_list_destroy};
 
-    for (int32_t partition = 0; auto timestamp : timestamps)
+    for (size_t i{0}; auto partitionTimestamp : partitionTimestamps)
```
Comment (chenziliang) [inline src/KafkaLog/KafkaWALCommon.cpp]: Follow the convention : i = 0;

**Review by chenziliang — CHANGES_REQUESTED** [review]

File: src/Core/Settings.h [hunk 5]
```diff
@@ -784,7 +789,11 @@ static constexpr UInt64 operator""_GiB(unsigned long long value)
     M(UInt64, max_windows, 1000, "Maximum number of streaming windows in one streaming query.", 0) \
     M(UInt64, fetch_buffer_size, 4 * 1024 * 1024, "Fetch buffer per query", 0) \
     M(UInt64, record_consume_batch_count, 1000, "Maximum number for consuming records at once", 0) \
-    M(Int64, record_consume_timeout, 100, "Timeout of consuming record", 0) \
+    M(Int64, record_consume_timeout_ms, 100, "Timeout of consuming record", 0) \
+    M(Int64, kafka_fetch_wait_max_ms, 100, "When reading from Kafka, max wait time", 0) \
```
Comment (chenziliang) [inline src/Core/Settings.h:793]: We will propagate these tuning knobs in a future PR

**Comment by jovezhong** [comment]
(Jove Github Bot) added it to the current sprint.

**Comment by jovezhong** [comment]
(Jove Github Bot) moved this ticket out of the GitHub project(up to 1200 tickets for one project).

**Pull request merged**